src/datasource.ts

```diff
@@ -17,4 +38,157 @@ export class DataSource extends DataSourceWithBackend<RedshiftQuery, RedshiftDat
 
   applyTemplateVariables = (query: RedshiftQuery, scopedVars: ScopedVars) =>
     applySQLTemplateVariables(query, scopedVars, getTemplateSrv);
+
+  query(request: DataQueryRequest<RedshiftQuery>): Observable<DataQueryResponse> {
+    if (config.featureToggles.redshiftAsyncQueryDataSupport) {
+      const targets = request.targets.filter(this.filterQuery);
+      if (!targets.length) {
+        return of({ data: [] });
+      }
+      const all: Array<Observable<DataQueryResponse>> = [];
+      for (let target of targets) {
+        if (target.hide) {
+          continue;
+        }
+        all.push(this.doSingle(target, request));
+      }
+      return merge(...all);
+    } else {
+      return super.query(request);
+    }
+  }
+
+  storeQuery(target: RedshiftQuery, queryInfo: RedshiftRunningQueryInfo) {
+    const key = JSON.stringify(target);
+    const existingQueryInfo = this.runningQueries[key] || {};
+    this.runningQueries[key] = { ...existingQueryInfo, ...queryInfo };
+  }
+
+  getQuery(target: RedshiftQuery) {
+    const key = JSON.stringify(target);
+    return this.runningQueries[key] || {};
+  }
+
+  removeQuery(target: RedshiftQuery) {
+    const key = JSON.stringify(target);
+    delete this.runningQueries[key];
+  }
+
+  doSingle(target: RedshiftQuery, request: DataQueryRequest<RedshiftQuery>): Observable<DataQueryResponse> {
+    let queryID: string | undefined = undefined;
+    let status: string | undefined = undefined;
+    let allData: DataFrame[] = [];
+
+    return getRequestLooper(
+      { ...request, targets: [target], requestId: `aws_ts_${requestCounter++}` },
+      {
+        /**
+         * Additional query to execute if the current query is still in a running state
+         */
+        getNextQuery: (rsp: DataQueryResponse) => {
+          if (rsp.data?.length) {
+            const first: DataFrame = rsp.data[0];
+            const meta = first.meta?.custom;
+
+            if (isRedshiftCustomMeta(meta) && isRunning(meta.status)) {
```

Comment:
`isRedshiftCustomMeta` is just a type guard added to avoid using `as RedshiftCustomMeta` here. Doesn't seem like we can type the meta property since `DataQueryResponse` isn't a generic.